craft_application/services/service_factory.py

```diff
@@ -62,6 +91,68 @@ class ServiceFactory:
 
     def __post_init__(self) -> None:
         self._service_kwargs: dict[str, dict[str, Any]] = {}
+        self._services: dict[str, services.AppService] = {}
```

Comment:
You can use `dataclasses` magic to handle those attributes as well, using `Field(init=False, default_factory=dict)`
```suggestion
        _service_kwargs: dict[str, dict[str, Any]] = Field(init=False, default_factory=dict)
        _services: dict[str, services.AppService] = Field(init=False, default_factory=dict)
```

Reply:
That makes those members of the Dataclass though, which I don't want. (In fact at some point I want to make this no longer a Dataclass, but that's backwards incompatible.)